python/revng/internal/api/synchronizer.py

```diff
@@ -0,0 +1,300 @@
+                    )
+                )
+
+            if not gather_bools(queue, "Failed to save"):
+                return False
+
+            # Before writing the index file proper, wait for the previous event to finish
+            if last_save_event is not None:
+                last_save_event.wait()
+
+            # Generate the `index.yml` file and upload it to S3
+            new_index = {k: v.s3_filename for k, v in new_file_list.items()}
+            with NamedTemporaryFile("w") as temp_index_file:
+                yaml.safe_dump(new_index, temp_index_file)
+                result = self._upload_file(Path(temp_index_file.name), "index.yml", "None")
+                if not result:
+                    return False
+
+            # Replace last_file_list with the new one
+            with self.last_file_list_lock:
+                self.last_file_list = new_file_list
+            return True
+
+    @staticmethod
+    def _generate_filename(filename: str) -> str:
+        filename_path = Path(filename)
+        return str(filename_path.parent / f"{uuid4()}-{filename_path.name}")
+
+    def set_credentials(self, credentials: str):
+        access_key, secret_key = credentials.split(":", 1)
+        self.botocore_session.set_credentials(access_key, secret_key)
+
+    def _get_key(self, name: str) -> str:
+        return os.path.join(self.path, name) if name != "" else self.path
+
+    def _download_file(self, key: str, destination: Path) -> bool:
+        destination.parent.mkdir(parents=True, exist_ok=True)
+        key = self._get_key(key)
+        try:
+            self.client.download_file(
+                Key=key,
+                Bucket=self.bucket,
+                Filename=str(destination.resolve()),
+            )
+            return True
+        except self.client.exceptions.NoSuchKey:
+            return False
+
+    def _read_file(self, name: str) -> bytes | None:
+        key = self._get_key(name)
+        try:
+            object_request = self.client.get_object(Bucket=self.bucket, Key=key)
+        except self.client.exceptions.NoSuchKey:
+            return None
+
+        return cast(bytes, object_request["Body"].read())
+
+    def _get_index(self) -> dict[str, str]:
+        data = self._read_file("index.yml")
+        if data is None:
+            return {}
+        return yaml.safe_load(data.decode("utf-8"))
+
+    def _upload_file(self, source: Path, name: str, encoding: str) -> bool:
+        key = self._get_key(name)
+        extra_args = {"ACL": "private"}
+        if encoding == "Gzip":
+            extra_args["ContentEncoding"] = "gzip"
+        try:
+            self.client.upload_file(
+                Filename=str(source.resolve()),
+                Key=key,
+                Bucket=self.bucket,
+                ExtraArgs=extra_args,
+            )
+            return True
+        except boto3.exceptions.S3UploadFailedError:
+            return False
```

Comment:
We are suppressing the error message entirely here, at least print it.

Reply:
Added print